File: scripts/download_images.py
#!/usr/bin/env python3
"""
Image Downloader for Dark Souls Wiki
Python version for downloading images from various sources
"""

import os
import sys
import json
import requests
from urllib.parse import urlparse
from pathlib import Path
import time

# Configuration
BASE_DIR = Path(__file__).parent.parent / 'assets' / 'images'
HEADERS = {
    'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36'
}

# Image URLs - Using direct URLs from various sources
IMAGE_URLS = {
    'weapons': {
        'zweihander': [
            'https://darksouls.wiki.fextralife.com/file/Dark-Souls/zweihander.png',
            'http://darksouls.wikidot.com/local--files/ultra-greatswords/zweihander.png'
        ],
        'claymore': [
            'https://darksouls.wiki.fextralife.com/file/Dark-Souls/claymore.png',
            'http://darksouls.wikidot.com/local--files/greatswords/claymore.png'
        ],
        'uchigatana': [
            'https://darksouls.wiki.fextralife.com/file/Dark-Souls/uchigatana-lg.png',
            'http://darksouls.wikidot.com/local--files/katanas/uchigatana.png'
        ],
        'black-knight-sword': [
            'https://darksouls.wiki.fextralife.com/file/Dark-Souls/black_knight_sword.png'
        ],
        'moonlight-greatsword': [
            'https://darksouls.wiki.fextralife.com/file/Dark-Souls/moonlight_greatsword.png'
        ],
        'estoc': [
            'https://darksouls.wiki.fextralife.com/file/Dark-Souls/estoc.png'
        ],
        'longsword': [
            'https://darksouls.wiki.fextralife.com/file/Dark-Souls/longsword.png'
        ],
        'dark-hand': [
            'https://darksouls.wiki.fextralife.com/file/Dark-Souls/dark_hand.png'
        ]
    },
    'bosses': {
        'asylum-demon': [
            'https://darksouls.wiki.fextralife.com/file/Dark-Souls/asylum_demon.jpg'
        ],
        'bell-gargoyles': [
            'https://darksouls.wiki.fextralife.com/file/Dark-Souls/bell_gargoyle.jpg'
        ],
        'taurus-demon': [
            'https://darksouls.wiki.fextralife.com/file/Dark-Souls/taurus_demon.jpg'
        ]
    },
    'areas': {
        'firelink-shrine': [
            'https://darksouls.wiki.fextralife.com/file/Dark-Souls/firelink_shrine.jpg'
        ],
        'undead-burg': [
            'https://darksouls.wiki.fextralife.com/file/Dark-Souls/undead_burg.jpg'
        ],
        'anor-londo': [
            'https://darksouls.wiki.fextralife.com/file/Dark-Souls/anor_londo.jpg'
        ],
        'blighttown': [
            'https://darksouls.wiki.fextralife.com/file/Dark-Souls/blighttown.jpg'
        ]
    },
    'npcs': {
        'solaire': [
            'https://darksouls.wiki.fextralife.com/file/Dark-Souls/solaire_of_astora.jpg'
        ],
        'siegmeyer': [
            'https://darksouls.wiki.fextralife.com/file/Dark-Souls/siegmeyer_of_catarina.jpg'
        ],
        'andre': [
            'https://darksouls.wiki.fextralife.com/file/Dark-Souls/andre_of_astora.jpg'
        ]
    },
    'items': {
        'estus-flask': [
            'https://darksouls.wiki.fextralife.com/file/Dark-Souls/estus_flask.png'
        ],
        'ring-of-favor': [
            'https://darksouls.wiki.fextralife.com/file/Dark-Souls/ring_of_favor_and_protection.png'
        ]
    }
}

class ImageDownloader:
    def __init__(self):
        self.session = requests.Session()
        self.session.headers.update(HEADERS)
        self.downloaded = 0
        self.failed = 0
        self.total = 0
        
    def download_image(self, url, filepath):
        """Download a single image from URL to filepath"""
        try:
            response = self.session.get(url, timeout=30, stream=True)
            response.raise_for_status()
            
            # Write image to file
            with open(filepath, 'wb') as f:
                for chunk in response.iter_content(chunk_size=8192):
                    if chunk:
                        f.write(chunk)
            
            return True
            
        except requests.exceptions.RequestException as e:
            print(f"  Error: {str(e)}")
            return False
    
    def get_extension(self, url):
        """Get file extension from URL"""
        parsed = urlparse(url)
        ext = os.path.splitext(parsed.path)[1]
        return ext if ext else '.jpg'
    
    def download_category(self, category, items):
        """Download all images in a category"""
        category_dir = BASE_DIR / category
        category_dir.mkdir(parents=True, exist_ok=True)
        
        print(f"\nDownloading {category}...")
        print("-" * 50)
        
        for item_id, urls in items.items():
            self.total += 1
            success = False
            
            # Try each URL until one succeeds
            for i, url in enumerate(urls):
                ext = self.get_extension(url)
                filename = f"{item_id}{ext}"
                filepath = category_dir / filename
                
                print(f"Downloading {item_id}...", end='')
                
                if self.download_image(url, filepath):
                    print(f" ✓ Success")
                    self.downloaded += 1
                    success = True
                    break
                else:
                    if i < len(urls) - 1:
                        print(f"\n  Trying alternative source {i+2}...", end='')
                    else:
                        print(f" ✗ Failed")
            
            if not success:
                self.failed += 1
            
            # Small delay to be polite to servers
            time.sleep(0.5)
    
    def download_all(self):
        """Download all images"""
        print("Dark Souls Wiki Image Downloader (Python)")
        print("=" * 50)
        
        # Download each category
        for category, items in IMAGE_URLS.items():
            self.download_category(category, items)
        
        # Summary
        print("\n" + "=" * 50)
        print("Download Summary:")
        print(f"Total: {self.total}")
        print(f"Downloaded: {self.downloaded}")
        print(f"Failed: {self.failed}")
        print("=" * 50)
        
        if self.failed > 0:
            print("\nNote: Some images failed to download.")
            print("You may need to:")
            print("1. Check your internet connection")
            print("2. Try running the script again")
            print("3. Download failed images manually")
    
    def generate_url_list(self):
        """Generate a JSON file with all image URLs for reference"""
        output_file = BASE_DIR.parent.parent / 'docs' / 'image-urls.json'
        
        with open(output_file, 'w') as f:
            json.dump(IMAGE_URLS, f, indent=2)
        
        print(f"\nImage URL list saved to: {output_file}")

def main():
    """Main function"""
    downloader = ImageDownloader()
    
    if len(sys.argv) > 1 and sys.argv[1] == '--urls':
        downloader.generate_url_list()
    else:
        try:
            downloader.download_all()
        except KeyboardInterrupt:
            print("\n\nDownload interrupted by user")
        except Exception as e:
            print(f"\n\nError: {str(e)}")

if __name__ == "__main__":
    main()

File: scripts/download-images.js
#!/usr/bin/env node

/**
 * Image Downloader for Dark Souls Wiki
 * Downloads images from various sources for the wiki
 */

const https = require('https');
const http = require('http');
const fs = require('fs');
const path = require('path');
const { URL } = require('url');

// Image URLs from various sources (using placeholder URLs - replace with actual Dark Souls image URLs)
const imageUrls = {
    weapons: {
        'zweihander': 'https://static.wikia.nocookie.net/darksouls/images/6/67/Zweihander.png',
        'claymore': 'https://static.wikia.nocookie.net/darksouls/images/9/91/Claymore.png',
        'uchigatana': 'https://static.wikia.nocookie.net/darksouls/images/5/54/Uchigatana.png',
        'black-knight-sword': 'https://static.wikia.nocookie.net/darksouls/images/5/57/Black_knight_sword.png',
        'moonlight-greatsword': 'https://static.wikia.nocookie.net/darksouls/images/9/90/Moonlight_Greatsword.png',
        'estoc': 'https://static.wikia.nocookie.net/darksouls/images/8/8d/Estoc.png',
        'longsword': 'https://static.wikia.nocookie.net/darksouls/images/7/74/Long_Sword.png',
        'dark-hand': 'https://static.wikia.nocookie.net/darksouls/images/2/28/Dark_Hand.png'
    },
    bosses: {
        'asylum-demon': 'https://static.wikia.nocookie.net/darksouls/images/4/44/Asylum_Demon.jpg',
        'bell-gargoyles': 'https://static.wikia.nocookie.net/darksouls/images/3/3e/Bell_Gargoyle.jpg',
        'taurus-demon': 'https://static.wikia.nocookie.net/darksouls/images/6/6f/Taurus_Demon.jpg'
    },
    areas: {
        'firelink-shrine': 'https://static.wikia.nocookie.net/darksouls/images/e/ef/Firelink_Shrine.jpg',
        'undead-burg': 'https://static.wikia.nocookie.net/darksouls/images/f/fd/Undead_Burg.jpg',
        'anor-londo': 'https://static.wikia.nocookie.net/darksouls/images/6/6a/Anor_Londo.jpg',
        'blighttown': 'https://static.wikia.nocookie.net/darksouls/images/3/3e/Blighttown.jpg'
    },
    npcs: {
        'solaire': 'https://static.wikia.nocookie.net/darksouls/images/4/44/Solaire.jpg',
        'siegmeyer': 'https://static.wikia.nocookie.net/darksouls/images/0/05/Siegmeyer.jpg',
        'andre': 'https://static.wikia.nocookie.net/darksouls/images/a/a5/Andre.jpg'
    },
    items: {
        'estus-flask': 'https://static.wikia.nocookie.net/darksouls/images/f/fd/Estus_Flask.png',
        'ring-of-favor': 'https://static.wikia.nocookie.net/darksouls/images/9/9c/Ring_of_favor_and_protection.png'
    }
};

// Alternative sources if primary fails
const alternativeUrls = {
    weapons: {
        'zweihander': 'https://darksouls.wiki.fextralife.com/file/Dark-Souls/zweihander.png',
        'claymore': 'https://darksouls.wiki.fextralife.com/file/Dark-Souls/claymore.png',
        'uchigatana': 'https://darksouls.wiki.fextralife.com/file/Dark-Souls/uchigatana-lg.png'
    }
};

class ImageDownloader {
    constructor() {
        this.baseDir = path.join(__dirname, '..', 'assets', 'images');
        this.downloaded = 0;
        this.failed = 0;
        this.total = 0;
    }

    async downloadImage(url, filepath) {
        return new Promise((resolve, reject) => {
            const protocol = url.startsWith('https') ? https : http;
            const file = fs.createWriteStream(filepath);

            const request = protocol.get(url, {
                headers: {
                    'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36'
                }
            }, (response) => {
                if (response.statusCode === 200) {
                    response.pipe(file);
                    file.on('finish', () => {
                        file.close();
                        resolve();
                    });
                } else if (response.statusCode === 301 || response.statusCode === 302) {
                    // Handle redirects
                    file.close();
                    fs.unlinkSync(filepath);
                    this.downloadImage(response.headers.location, filepath)
                        .then(resolve)
                        .catch(reject);
                } else {
                    file.close();
                    fs.unlinkSync(filepath);
                    reject(new Error(`Failed to download: ${response.statusCode}`));
                }
            });

            request.on('error', (err) => {
                file.close();
                fs.unlinkSync(filepath);
                reject(err);
            });

            file.on('error', (err) => {
                fs.unlinkSync(filepath);
                reject(err);
            });
        });
    }

    async downloadCategory(category, urls) {
        const categoryDir = path.join(this.baseDir, category);
        
        // Create directory if it doesn't exist
        if (!fs.existsSync(categoryDir)) {
            fs.mkdirSync(categoryDir, { recursive: true });
        }

        console.log(`\nDownloading ${category}...`);
        console.log('─'.repeat(50));

        for (const [id, url] of Object.entries(urls)) {
            this.total++;
            const extension = path.extname(new URL(url).pathname) || '.jpg';
            const filename = `${id}${extension}`;
            const filepath = path.join(categoryDir, filename);

            try {
                console.log(`Downloading ${id}...`);
                await this.downloadImage(url, filepath);
                console.log(`✓ ${id} downloaded successfully`);
                this.downloaded++;
            } catch (error) {
                console.error(`✗ Failed to download ${id}: ${error.message}`);
                
                // Try alternative URL if available
                if (alternativeUrls[category] && alternativeUrls[category][id]) {
                    console.log(`  Trying alternative source...`);
                    try {
                        await this.downloadImage(alternativeUrls[category][id], filepath);
                        console.log(`  ✓ ${id} downloaded from alternative source`);
                        this.downloaded++;
                    } catch (altError) {
                        console.error(`  ✗ Alternative also failed: ${altError.message}`);
                        this.failed++;
                    }
                } else {
                    this.failed++;
                }
            }
        }
    }

    async downloadAll() {
        console.log('Dark Souls Wiki Image Downloader');
        console.log('================================\n');

        // Ensure base directory exists
        if (!fs.existsSync(this.baseDir)) {
            fs.mkdirSync(this.baseDir, { recursive: true });
        }

        // Download each category
        for (const [category, urls] of Object.entries(imageUrls)) {
            await this.downloadCategory(category, urls);
        }

        // Summary
        console.log('\n' + '='.repeat(50));
        console.log('Download Summary:');
        console.log(`Total: ${this.total}`);
        console.log(`Downloaded: ${this.downloaded}`);
        console.log(`Failed: ${this.failed}`);
        console.log('='.repeat(50));

        if (this.failed > 0) {
            console.log('\nNote: Some images failed to download.');
            console.log('You may need to:');
            console.log('1. Check your internet connection');
            console.log('2. Update the URLs in the script');
            console.log('3. Download failed images manually');
        }
    }
}

// Run the downloader
if (require.main === module) {
    const downloader = new ImageDownloader();
    downloader.downloadAll().catch(console.error);
}

module.exports = ImageDownloader;